proto/docs/arbitration.md

```diff
@@ -1,74 +1,125 @@
+      connection, the server should terminate the stream by returning an
+      `INVALID_ARGUMENT` error.
+    - Otherwise, the server updates the `election_id` for this client. If this makes
+      the client the new master, the server will also accept the given 'role.config`.
+  - Otherwise (i.e. `role.id` is different from current one), the server moves the 
+    client to the new role, and processes the update as described above. The server
+    accepts the given `role.config` only if the client becomes master for this
+    role based on the given `election_id`.
+
+### `MasterArbitrationUpdate` from server to client
+
+- If the mastership changed following a `MasterArbitrationUpdate` client
+  request, all existing clients for that role are sent a `MasterArbitrationUpdate`
+  message by the server. The `device_id` must be set. The 'role' and 'election_id'
+  fields must be set to reflect the master client's view. The `status` field code
+  must be `OK` for the master, and `ALREADY_EXISTS` for all the slave clients.
+
+- Otherwise, only the client which sent the request is notified, e.g. a new
+  connection, or a change in `role` and/or `election_id`. The `device_id` must be
+  set. The 'role' and 'election_id' fields must be set to reflect the master
+  client's view. The `status` field code must be `OK` if this client is the master,
+  and `ALREADY_EXISTS` if it is a slave.
+
+### Packet-In
 
 Only the master receives `PacketIn` messages. If there is no master (i.e. no
 client connections), packets are dropped.
 
-### Packet-out
+### Packet-Out
 
 Only the master can send `PacketOut` messages. In all other cases, the message
 is dropped.
 
-### Write requests
+### Write RPC
+
+The server will only perform write requests from the master client (for a given
+role based on `role_id` and `election_id` in the `WriteRequest`). If a slave tries
+to issue a `Write`, error code `PERMISSION_DENIED` will be returned.
 
-Only the master can send `WriteRequest` messages. If a slave tries to issue a
-`Write`, error code `PERMISSION_DENIED` is returned. The master is indetified by
-using the `election_id` field. Note that all clients are trusted; we do not
-support the case where a client is using an incorrect `election_id` in his
-`WriteRequest` messages.
+The `WriteRequest` will also be validated against the role config corresponding to
+the given `role_id`. If a given Write `update` fails this validation, error code
+`PERMISSION_DENIED` will be returned for that update.
 
-Note that for the sake of debugging and testing, we accept `WriteRequest`
+NOTE: All clients are trusted; we do not support the case where a client is using
+an incorrect `role_id` and `election_id` in its `WriteRequest` messages.
+
+NOTE: For the sake of debugging and testing, we accept `WriteRequest`
 messages when there is no master (i.e. no client connections) and the
-`election_id` field is not set.
+`role_id` and `election_id` fields are not set.
```

Comment:
Apart from the debugging case, I believe we should require `role_id` to be default (0) is this controller instance had an unset `role` field in its `MasterArbitrationUpdate` message.

Reply:
Called this out in both 'Write RPC' and 'SetForwardingPipelineConfig RPC' sections.